README.md
# Adaptation of Vtree
#### Avik Kadakia and Eugene Stark

# Introduction

In this project I have updated an old piece of
software, making sure it compiles, and that it works properly.

Maintaining old code is a chore and an often hated part of software
engineering. It is definitely one of the aspects which are seldom
discussed or thought about by aspiring computer science students.
However, it is prevalent throughout industry and a worthwhile skill to
learn.  Of course, this project did not give you a remotely
realistic experience in maintaining legacy code or code left behind by
previous engineers but it still provides a small taste of what the
experience may be like.  I took on the role of an engineer
whose supervisor has asked me to correct all the errors in the
program, plus add additional functionality.

## The Existing Program

The goal is extend an old program called `vtree`,
which was posted to Usenet sometime between March and May, 1989.
It was written by Jonathan Bayer of Intelligent Software Products, Inc.,
and it ran under the following operating systems:
4.2BSD, 4.3BSD, System V, System III, SCO Xenix 386, and SCO Xenix 286.

The purpose of the `vtree` program is to traverse a portion of the Unix
filesystem hierarchy, collect information on the number of inodes
(basically, files and directories) and the amount of disk space used,
and to display a representation of the hierarchy in a "visual" format,
possibly with summary statistics.  The functionality is similar to,
but less refined than, the more modern program `tree`, which is installed
on a Linux Mint system.

In the default mode of execution (which occurs when you invoke the program
just using `bin/vtree` with no arguments), the program traverses the filesystem
hierarchy rooted at the current working directory and displays the name of
each file and directory and the number of K-bytes of disk space occupied by
each of them.  Various options can be supplied to modify the behavior of
the program, as discussed in more detail below.

Here is the structure of the base code:

<pre>
.
├── .gitlab-ci.yml
└── hw2
    ├── doc
    │   ├── README
    │   └── vtree.1
    ├── include
    │   ├── customize.h
    │   ├── debug.h
    │   ├── hash.h
    │   └── patchlevel.h
    ├── Makefile
    ├── src
    │   ├── direct.c
    │   ├── hash.c
    │   ├── main.c
    │   └── vtree.c
    └── tests
        ├── hw2_tests.c
        └── rsrc
            ├── duplicates_test.err
            ├── duplicates_test.in
            ├── duplicates_test.out
            ├── getopt_test.err
            ├── getopt_test.in
            ├── getopt_test.out
            ├── inode_test.err
            ├── inode_test.in
            ├── inode_test.out
            ├── option_error_test.err
            ├── option_error_test.in
            ├── option_error_test.out
            ├── quick_test.err
            ├── quick_test.in
            ├── quick_test.out
            ├── sort_test.err
            ├── sort_test.in
            ├── sort_test.out
            ├── test_tree
            │   ├── A
            │   ├── B
            │   ├── C
            │   ├── D
            │   └── S
            │       ├── a
            │       │   └── 4K
            │       ├── b
            │       │   └── 8K
            │       ├── c
            │       │   └── 12K
            │       ├── d
            │       │   └── 16K
            │       └── E
            ├── valgrind_leak_test.err
            ├── valgrind_leak_test.in
            ├── valgrind_leak_test.out
            ├── valgrind_uninitialized_test.err
            ├── valgrind_uninitialized_test.in
            └── valgrind_uninitialized_test.out
</pre>

The `doc` directory included with the assignment basecode contains the
original documentation files (a `README` and a Unix-style `man` page) that
were distributed with the program.  You can format and read the `man` page
using the following command:

<pre>
nroff -man doc/vtree.1 | less
</pre>

  > Since time immemorial, Unix `man` pages have been written in
  > a typesetting language called `roff` (short for "run-off").
  > The `nroff` program processes `roff` source and produces text formatted
  > for reading in a terminal window.

The source files for the `vtree` application are in the `src` directory, with associated
header files in the `include` directory.  There are only three source files:
`vtree.c` which is the main part of the code, `hash.c`, which implements a hash table
for the purpose of avoiding double-counting of files that can be accessed via more than
one path, and `direct.c`, which contains some implementations of directory-reading functions
that were required on Unix variants  that did not have the `readdir` function supported
by BSD and System V.
This function, which is now part of the POSIX standard, is supported by modern Unix-like
systems including Linux, so we will not need to actually use the functions in this file.
However, as our objective is to make the program compile and run on Linux without affecting
its ability to compile and run on the other supported systems (though we won't really be able
to test this), we do not want to delete this file entirely.

The code contains sections that are conditionalized on various preprocessor symbols
being defined:

- The operating system on which the program is being compiled is specified by defining
  one of the symbols `BSD`, `SYS_V`, `SYS_III`, `SCO_XENIX`.  I added support
  for Linux, conditionalized on `LINUX`.

- Some other variations can be obtained by defining zero or more of the symbols
  `MEMORY_BASED`, `HSTATS`, and `LSTAT`, and `ONEPERLINE`:

	- Compiling with `MEMORY_BASED` defined causes code to be used that builds a data structure
      in memory to represent the directory hierarchy being traversed.  This avoids having
	  to traverse each directory twice and it also permits directory entries to be presented in
	  sorted order.
	- Compiling with `HSTATS` defined causes the program to print out statistics about the
      hash table.
	- Compiling with `LSTAT` defined supposedly enables code to avoid following symbolic links,
      but it seems that it doesn't function reasonably.  You will correct it.
    - I didn't manage to understand clearly what `ONEPERLINE` was supposed to do, so you can
	  just ignore it.

The `Makefile` provides you with the option of selecting various compilation flags from
the command line.  For example:
```
$ make OS=BSD
```
will define the `BSD` preprocessor symbol instead of the `LINUX` symbol, which is the default.
Also:
```
make OPTIONS="-DMEMORY_BASED -DHSTATS"
```
will define both the `MEMORY_BASED` and `HSTATS` preprocessor symbols.

The `tests` directory contains C source code (in file `hw2_tests.c`) for some
Criterion tests Eugene Stark have supplied.
The subdirectory `tests/rsrc` contains data files that are used by the tests.

# Present Command Line Arguments

The command line arguments and expected operation of the program are described
by the following "Usage" message, which is printed within `vtree_main()` in `vtree.c`:

```
bin/vtree: [ -d ] [ -h # ] [ -i ] [ -o ] [ -s ] [ -q ] [ -v ] [ -V ]
	-d	count duplicate inodes
	-f	floating column widths
	-h #	height of tree to look at
	-i	count inodes
	-o	sort directories before processing
	-s	include subdirectories not shown due to -h option
	-t	totals at the end
	-q	quick display, no counts
	-v	visual display
	-V	show current version
		(2 Vs shows specified options)
```

- The `-d` option causes `vtree` not to use a hash table to check whether a particular inode
(file or directory) has already been encountered and its size tallied.  This might result in
the program over-estimating the total space used.

- I don't know what the `-f` option is really supposed to do.  Ignore it.

- The `-h` option causes `vtree` to limit its traversal of subdirectories to the specified
"height" (personally, I would call it "depth").

- The `-i` option asks `vtree` to count and report the numbers of inodes (files or directories),
in addition to the amount of disk space used.

- The `-o` option causes `vtree` to sort directory entries in lexicographic order before printing
them.  This option will only work if the program has been compiled with `MEMORY_BASED` defined.

- The `-s` option causes `vtree` to descend into subdirectories and calculate their size, etc.,
even if these subdirectories are not shown due to their being beyond the horizon specified with `-h`.

- The `-t` option causes `vtree` to show summary totals at the end of the tree printout.

- The `-q` option just displays the directories, without size or inode information.

- The `-v` option enables a "visual" display, which perhaps shows the tree structure more clearly.

- The `-V` option displays the program version and whether or not the program was compiled
with the `MEMORY_BASED` compilation option.  The `-VV` option also shows information about
the command-line options that were supplied when the program was run.

# Part 2: Added Features

The original program uses `getopt()` to do options processing, but it only understands
traditional, single-character options.
For this part of the assignment, I rewrote the options processing so that it uses
`getopt_long()` instead of `getopt()`, and it understands the following alternative
forms for the various options (as well as the original short forms):

   - `--duplicates` as equivalent to `-d`
   - `--floating-column-widths` as equivalent to `-f`
   - `--height` as equivalent to `-h`
   - `--inodes` as equivalent to `-i`
   - `--sort-directories` as equivalent to `-o`
   - `--totals` as equivalent to `-t`
   - `--quick-display` as equivalent to `-q`
   - `--visual-display` as equivalent to `-v`
   - `--version` as equivalent to `-V`

In addition, I also added the following new option:

   - `--no-follow-symlinks` as equivalent to `-l`


This option is only be available if the program was compiled with the `LSTAT`
preprocessor variable defined.  It's effect is to cause the program **not** to
follow symbolic links when they are encountered, as opposed to the default behavior,
which is to follow symbolic links.  In addition, the existence of the option will
be reported in the `Usage` message.
If the program was not compiled with `LSTAT` defined, then this option is illegal
and is not be mentioned in the `Usage` message.

Finally, arranged for the `-o` (sorted output) option to be available only if the program
has been compiled with the `MEMORY_BASED` symbol defined.  If this symbol has not defined,
then the program does not accept the `-o` option nor mention it in the "Usage" message.

Changes.txt
Changes Made:

    customize.c:
            direct.c
                Added the ifdef statement for LINUX
                Added LINUX in the if statement as an || option on line 54
            // vtree.c
                // For all NAME functions, changed ((x).d_name) -> ((x)->d_name)


    direct.c:
        Moved opendir method in the ifdef for SYS_III
        Changed dp -> dirp in the readdir() method

    hash.h:
        Added:
            int h_enter(dev_t dev, ino_t ino)
            void h_stats()
        Created an array: memory_array and variable: positions

    hash.c:
        Change return type to void for the 3 functions - void char *malloc(), void char *realloc(), void char *calloc()
        h_enter return type is int
        Incremented positions everytime calloc, malloc, or realloc is called and added to the position in the memory_array

    vtree.c:
        In RD_List pointer, READ entry -> READ *entry
        Commented READ *readdir() and OPEN *opendir()
        170:7 : tmp_entry -> *tmp_entry
        included <stdlib.h>
        341, 378, 406: &tmp_RD->entry -> tmp_RD->entry
        292: (NAME(*file), "..") -> (NAME(file), "..")
        Renamed dup -> duplicate
        down method has return type void
        176: READ *tmp_entry -> READ tmp_entry
        Added on the top:
            static int chk_4_dir(char *path)
            static void get_data(char *path, int cont)
            static int is_directory(char *path)
        Changed the malloc(struct RD_List *) -> malloc(struct RD_List )
        Made head and tail equal to null
        Added free:
            Line 419, 504
        Cleared the memory_array at the end of vtree_main()
        Changed 714: printf("Total inodes: %d\n", inodes) -> printf("Total inodes: %d\n", total_inodes);
        Changes getopt to getopt_long

src/direct.c
/* direct.c
  
   SCCS ID	@(#)direct.c	1.6	7/9/87
  
 *
 *	My own substitution for the berkeley reading routines,
 *	for use on System III machines that don't have any other
 *	alternative.
 */

#include "customize.h"

#define NAMELENGTH 14
#ifdef SYS_III
FILE *opendir(name) { return (fopen(name, "r")); }

/*
  * Read a directory, returning the next (non-empty) slot. 
  */

READ *readdir(OPEN *dirp)
{
    static READ direct;

    /* This read depends on direct being similar to dir_entry. */

    while (fread(&direct, sizeof(struct dir_entry), 1, dirp) != 0)
    {
        direct.terminator = '\0';
        if (INO(direct) != 0)
            return &direct;
    };

    return (READ *)NULL;
}

struct dir_entry
{ /* What the system uses internally. */
    ino_t d_ino;
    char d_name[NAMELENGTH];
};

struct direct
{ /* What these routines return. */
    ino_t d_ino;
    char d_name[NAMELENGTH];
    char terminator;
};

#else
#define opendir(name) fopen(name, "r")
#endif
#define closedir(fp) fclose(fp)

include/customize.h
/* 
   This is the customizations file.  It changes our ideas of
   how to read directories.
*/

#define NAMELEN 512 /* max size of a full pathname */

#ifdef BSD
#include <sys/dir.h>
#define OPEN DIR
#define READ struct direct
#define NAME(x) ((x).d_name)
#endif

#ifdef SCO_XENIX
#include <sys/ndir.h>
#define OPEN DIR
#define READ struct direct
#define NAME(x) ((x).d_name)
#endif

#ifdef SYS_V
/* Customize this.  This is part of Doug Gwyn's package for */
/* reading directories.  If you've put this file somewhere */
/* else, edit the next line. */

#include <dirent.h>

#define OPEN struct direct
#define READ struct dirent
#define NAME(x) ((x).d_name)
#endif

#ifdef SYS_III
#define OPEN FILE
#define READ struct direct
#define NAME(x) ((x).d_name)
#define INO(x) ((x).d_ino)

#include "direct.c"

#endif

#ifdef LINUX

#include <dirent.h>

#define OPEN DIR
#define READ struct dirent // direct
#define NAME(x) ((x).d_name)

#endif

#if !(defined(BSD) || defined(LINUX) || !defined(SYS_V) || !defined(SYS_III) || !defined(SCO_XENIX))
"This is an Error"
#endif


src/vtree.c
/* vtree
  
   +=======================================+
   | This program is in the public domain. |
   +=======================================+
  
   This program shows the directory structure of a filesystem or 
   part of one.  It also shows the amount of space taken up by files
   in each subdirectory. 
  
   Call via
  
	vtree fn1 fn2 fn3 ...
  
   If any of the given filenames is a directory (the usual case),
   vtree will recursively descend into it, and the output line will
   reflect the accumulated totals for all files in the directory.
   
   This program is based upon "agef" written by David S. Hayes at the 
   Army Artificial Intelligence Center at the Pentagon.
   
   This program is dependent upon the new directory routines written by
   Douglas A. Gwyn at the US Army Ballistic Research Laboratory at the
   Aberdeen Proving Ground in Maryland.
*/
/*
** Patches were received from the following people:
**
**	1.	Mike Howard, (...!uunet!milhow1!how)
**		Mike's patches included changes to the Makefile to
**		customize vtree to SCO Xenix for the 286 as well as the
**		386.  He also added external definitions to hash.c
**
**	2.	Andrew Weeks, (...!uunet!mcvax!doc.ic.ac.uk!aw)
**		Andrew sent me diffs to make vtree work properly under BSD
**		He also pointed out that you will need one of the PD getopt
**		packages for BSD.
**
**	3.	Ralph Chapman, (...uunet!ihnp4!ihuxy!chapman)
**		Ralph sent me changes (not diffs unfortunately) to make
**		vtree work properly under the SYS_III option.  His changes
**		were in direct.c and vtree.c
**
**	4.	David Eckelkamp notified me of a bug when printing the
**		visual tree.  The bug occured when a directory name
**		was too long.  It caused vtree to mess up the tree
**		being printed.
*/

// My additions:

#include <unistd.h>
#include <stdlib.h>
#include "debug.h"
#include <getopt.h>

// Already Present

#include "patchlevel.h"

#include <ctype.h>
#include <sys/types.h>
#include <sys/stat.h>
#include <sys/param.h>
#include <stdio.h>
#include <string.h>

#include "customize.h"
#include "hash.h"

#ifdef SYS_III
#define rewinddir(fp) rewind(fp)
#endif

#define SAME 0		  /* for strcmp */
#define BLOCKSIZE 512 /* size of a disk block */

#define K(x) ((x + 1023) / 1024) /* convert stat(2) blocks into  \
								  * k's.  On my machine, a block \
								  * is 512 bytes. */

#define TRUE 1
#define FALSE 0
#define V_CHAR "|"  /*	Vertical character	*/
#define H_CHAR "-"  /*	Horizontal character	*/
#define A_CHAR ">"  /*	Arrow char		*/
#define T_CHAR "+"  /*	Tee char		*/
#define L_CHAR "\\" /*	L char, bottom of a branch	*/

#define MAX_COL_WIDTH 15
#define MAX_V_DEPTH 256 /* max depth for visual display */
#ifdef MEMORY_BASED
struct RD_list
{
	READ entry;
	struct RD_list *fptr;
	struct RD_list *bptr;
};
#endif

int indent = 0,   /* current indent */
	depth = 9999, /* max depth */
	cur_depth = 0,
	sum = FALSE,	   /* sum the subdirectories */
	duplicate = FALSE, /* use duplicate inodes */
	floating = FALSE,  /* floating column widths */
	sort = FALSE,
	cnt_inodes = FALSE, /* count inodes */
	quick = FALSE,		/* quick display */
	visual = FALSE,		/* visual display */
	version = 0,		/* = 1 display version, = 2 show options */
	sub_dirs[MAX_V_DEPTH],
	sub_dirs_indents[MAX_V_DEPTH];

struct stat stb; /* Normally not a good idea, but */
				 /* this structure is used through- */
				 /* out the program		   */

extern char *optarg; /* from getopt(3) */
extern int optind, opterr;

char *Program;			   /* our name */
short sw_follow_links = 1; /* follow symbolic links */
short sw_summary;		   /* print Grand Total line */

int total_inodes, inodes; /* inode count */
long total_sizes, sizes;  /* block count */

char topdir[NAMELEN]; /* our starting directory */

// Defining functions beforehand for vtree.c
static int chk_4_dir(char *path);
static void get_data(char *path, int cont);
static int is_directory(char *path);

/**
 * Find the last field of a string.
 * 
 * @param p Null-terminated string to scan
 * @param c Separator char, usually '/'
*/
char *lastfield(char *p, int c)
{
	char *r;

	r = p;
	while (*p) /* Find the last field of the name */
		if (*p++ == c)
			r = p;
	return r;
} /* lastfield */

/*
  * We ran into a subdirectory.  Go down into it, and read everything
  * in there. 
  */
int indented = FALSE;	/* These had to be global since they */
int last_indent = 0;	 /* determine what gets displayed during */
int last_subdir = FALSE; /* the visual display */

void down(char *subdir)
{
	OPEN *dp; /* stream from a directory */
	OPEN *opendir();
	READ *readdir();
	char cwd[NAMELEN], tmp[NAMELEN];
	char *sptr;
	READ *file; /* directory entry */
	int i, x;
	struct stat stb;

#ifdef MEMORY_BASED
	struct RD_list *head, *tail, *tmp_RD, *tmp1_RD; /* head and tail of directory list */
	struct RD_list sz;
	READ tmp_entry;
#endif

	if ((cur_depth == depth) && (!sum))
		return;

	/* display the tree */

	if (cur_depth < depth)
	{
		if (visual)
		{
			if (!indented)
			{
				for (i = 1; i < cur_depth; i++)
				{
					if (floating)
						x = sub_dirs_indents[i] + 1;
					else
						x = MAX_COL_WIDTH - 3;
					if (sub_dirs[i])
					{
						printf("%*s%s   ", x - 1, " ", V_CHAR);
					}
					else
						printf("%*s   ", x, " ");
				}
				if (cur_depth > 0)
				{
					if (floating)
						x = sub_dirs_indents[cur_depth] + 1;
					else
						x = MAX_COL_WIDTH - 3;
					if (sub_dirs[cur_depth] == 0)
					{
						printf("%*s%s%s%s ", x - 1, " ", L_CHAR, H_CHAR, A_CHAR);
						last_subdir = cur_depth;
					}
					else
						printf("%*s%s%s%s ", x - 1, " ", T_CHAR, H_CHAR, A_CHAR);
				}
			}
			else
			{
				if (!floating)
					for (i = 1; i < MAX_COL_WIDTH - last_indent - 3; i++)
						printf("%s", H_CHAR);
				printf("%s%s%s ", T_CHAR, H_CHAR, A_CHAR);
			}

	/* This is in case a subdir name is too big.  It is then displayed on
	** two lines, the first line is the full name, the second line is
	** truncated.  Any subdirs displayed for the current subdir will be
	** appended to the second line.  This keeps the columns in order
	*/

#ifndef ONEPERLINE
			if ((strlen(subdir) > MAX_COL_WIDTH - 3 && !floating))
			{
#else
			if ((strlen(subdir) > MAX_COL_WIDTH - 3 && !floating) || lastfield(subdir, '/') != subdir)
			{
#endif
				printf("%s\n", subdir);
				for (i = 1; i <= cur_depth; i++)
				{
					if (sub_dirs[i])
					{
						printf("%*s%s   ", MAX_COL_WIDTH - 4, " ", V_CHAR);
					}
					else
						printf("%*s   ", MAX_COL_WIDTH - 3, " ");
				}
				strcpy(tmp, lastfield(subdir, '/'));
				tmp[MAX_COL_WIDTH - 4] = 0;
				printf("%s", tmp);
#ifdef ONEPERLINE
				if (floating || strlen(tmp) < MAX_COL_WIDTH - 4)
					printf(" ");
#endif
				sub_dirs_indents[cur_depth + 1] = last_indent = strlen(tmp) + 1;
			}
			else
			{
				printf("%s", subdir);
				sub_dirs_indents[cur_depth + 1] = last_indent = strlen(subdir) + 1;
				if (floating || strlen(subdir) < MAX_COL_WIDTH - 4)
					printf(" ");
			}
			indented = TRUE;
		}
		else
			printf("%*s%s", indent, " ", subdir);
	}

	/* open subdirectory */

	if ((dp = opendir(subdir)) == NULL)
	{
		printf(" - can't read %s\n", subdir);
		indented = FALSE;
		return;
	}

	cur_depth++;
	indent += 3;

#ifdef BSD
	getcwd(cwd, sizeof(cwd)); /* remember where we are */
#else
	getcwd(cwd, sizeof(cwd)); /* remember where we are */
#endif
	chdir(subdir); /* go into subdir */

#ifdef MEMORY_BASED

	head = NULL;
	tail = NULL;

	for (file = readdir(dp); file != NULL; file = readdir(dp))
	{
		if ((!quick && !visual) || (strcmp(NAME(*file), "..") != SAME && strcmp(NAME(*file), ".") != SAME && chk_4_dir(NAME(*file))))
		{
			tmp_RD = (struct RD_list *)malloc(sizeof(struct RD_list));
			memcpy(&tmp_RD->entry, file, sizeof(tmp_entry));
			tmp_RD->bptr = head;
			tmp_RD->fptr = NULL;
			if (head == NULL)
				head = tmp_RD;
			else
				tail->fptr = tmp_RD;

			tail = tmp_RD;
		}
	}

	/* screwy, inefficient, bubble sort	*/
	/* but it works				*/
	if (sort)
	{
		tmp_RD = head;
		while (tmp_RD)
		{
			tmp1_RD = tmp_RD->fptr;
			while (tmp1_RD)
			{
				if ((strcasecmp(NAME(tmp_RD->entry), NAME(tmp1_RD->entry)) > 0) &&
					((strcasecmp(NAME(tmp_RD->entry), ".") != 0) ||
					 (strcasecmp(NAME(tmp1_RD->entry), ".") != 0) ||
					 (strcasecmp(NAME(tmp_RD->entry), "..") != 0) ||
					 (strcasecmp(NAME(tmp1_RD->entry), "..") != 0)))
				{
					/* swap the two */
					memcpy(&tmp_entry, &tmp_RD->entry, sizeof(tmp_entry));
					memcpy(&tmp_RD->entry, &tmp1_RD->entry, sizeof(tmp_entry));
					memcpy(&tmp1_RD->entry, &tmp_entry, sizeof(tmp_entry));
				}
				tmp1_RD = tmp1_RD->fptr;
			}
			tmp_RD = tmp_RD->fptr;
		}
	}

#endif

	if ((!quick) && (!visual))
	{

		/* accumulate total sizes and inodes in current directory */

#ifdef MEMORY_BASED
		tmp_RD = head;
		while (tmp_RD)
		{
			file = &tmp_RD->entry;
			tmp_RD = tmp_RD->fptr;

#else

		for (file = readdir(dp); file != NULL; file = readdir(dp))
		{
#endif
			if (strcmp(NAME(*file), "..") != SAME)
				get_data(NAME(*file), FALSE);
		}

		if (cur_depth < depth)
		{
			if (cnt_inodes)
				printf("   %d", inodes);
			printf(" : %ld\n", sizes);
			total_sizes += sizes;
			total_inodes += inodes;
			sizes = 0;
			inodes = 0;
		}
#ifndef MEMORY_BASED
		rewinddir(dp);
#endif
	}
	else if (!visual)
		printf("\n");

	if (visual)
	{

		/* count subdirectories */

#ifdef MEMORY_BASED
		tmp_RD = head;
		while (tmp_RD)
		{
			file = &tmp_RD->entry;
			tmp_RD = tmp_RD->fptr;
#else
		for (file = readdir(dp); file != NULL; file = readdir(dp))
		{
			if ((strcmp(NAME(*file), "..") != SAME) &&
				(strcmp(NAME(*file), ".") != SAME))
			{
				if (chk_4_dir(NAME(*file)))
				{
#endif
			sub_dirs[cur_depth]++;
#ifndef MEMORY_BASED
		}
	}
#endif
}
#ifndef MEMORY_BASED
rewinddir(dp);
#endif
}

/* go down into the subdirectory */

#ifdef MEMORY_BASED
tmp_RD = head;
while (tmp_RD)
{
	file = &tmp_RD->entry;
	tmp_RD = tmp_RD->fptr;
#else
			for (file = readdir(dp); file != NULL; file = readdir(dp))
			{
#endif
	if ((strcmp(NAME(*file), "..") != SAME) &&
		(strcmp(NAME(*file), ".") != SAME))
	{
		if (chk_4_dir(NAME(*file)))
			sub_dirs[cur_depth]--;
		get_data(NAME(*file), TRUE);
	}
}

if ((!quick) && (!visual))
{

	/* print totals */

	if (cur_depth == depth)
	{
		if (cnt_inodes)
			printf("   %d", inodes);
		printf(" : %ld\n", sizes);
		total_sizes += sizes;
		total_inodes += inodes;
		sizes = 0;
		inodes = 0;
	}
}

#ifdef MEMORY_BASED
/* free the allocated memory */
tmp_RD = head;
struct RD_list *temp;

while (head != NULL)
{
	temp = head->fptr;
	free(head);
	head = temp;
}

#endif

if (visual && indented)
{
	printf("\n");
	indented = FALSE;
	if (last_subdir >= cur_depth - 1)
	{
		for (i = 1; i < cur_depth; i++)
		{
			if (sub_dirs[i])
			{
				if (floating)
					printf("%*s%s   ", sub_dirs_indents[i], " ", V_CHAR);
				else
					printf("%*s%s   ", MAX_COL_WIDTH - 4, " ", V_CHAR);
			}
			else
			{
				if (floating)
					/*ZZZ*/ printf("%*s   ", sub_dirs_indents[i] + 1, " ");
				else
					printf("%*s   ", MAX_COL_WIDTH - 3, " ");
			}
		}
		printf("\n");
		last_subdir = FALSE;
	}
}
indent -= 3;
sub_dirs[cur_depth] = 0;
cur_depth--;

free(dp);
chdir(cwd); /* go back where we were */
// closedir(dp);

} /* down */

int chk_4_dir(char *path)
{
	if (is_directory(path))
		return TRUE;
	else
		return FALSE;

} /* chk_4_dir */

/* Is the specified path a directory ? */

int is_directory(char *path)
{

#ifdef LSTAT
	if (sw_follow_links)
		stat(path, &stb); /* follows symbolic links */
	else
		lstat(path, &stb); /* doesn't follow symbolic links */
#else
			stat(path, &stb);
#endif

	if ((stb.st_mode & S_IFMT) == S_IFDIR)
		return TRUE;
	else
		return FALSE;
} /* is_directory */

/*
  * Get the aged data on a file whose name is given.  If the file is a
  * directory, go down into it, and get the data from all files inside. 
  */

void get_data(char *path, int cont)
{
	int i;

	if (cont)
	{
		if (is_directory(path))
		{
			inodes++;
			sizes += stb.st_blocks / 2;
			down(path);
		}
	}
	else
	{
		if (is_directory(path))
			return;

		/* Don't do it again if we've already done it once. */

		if ((h_enter(stb.st_dev, stb.st_ino) == OLD) && (!duplicate))
			return;

		inodes++;
		sizes += stb.st_blocks / 2;
	}
} /* get_data */

int vtree_main(int argc, char *argv[])
{
	int i,
		j,
		err = FALSE;
	int option;
	int user_file_list_supplied = 0;

	Program = *argv; /* save our name for error messages */

	/* Pick up options from command line */

	struct option long_options[] = {
		{"duplicates", 0, 0, 'd'},
		{"floating-column-widths", 0, 0, 'f'},
		{"height", 1, 0, 'h'},
		{"inodes", 0, 0, 'i'},
		{"sort-directories", 0, 0, 'o'},
		{"totals", 0, 0, 't'},
		{"quick-display", 0, 0, 'q'},
		{"visual-display", 0, 0, 'v'},
		{"version", 0, 0, 'V'},
		{"no-follow-symlinks", 0, 0, 'l'},
		{0, 0, 0, 0}};

	int option_index = 0;
	while ((option = getopt_long(argc, argv, "dfh:ilostqvV", long_options, &option_index)) != EOF)
	{
		switch (option)
		{
		case 'f':
			floating = TRUE;
			break;
		case 'h':
			depth = atoi(optarg);
			while (*optarg)
			{
				if (!isdigit(*optarg))
				{
					err = TRUE;
					break;
				}
				optarg++;
			}
			break;
		case 'd':
			duplicate = TRUE;
			break;
		case 'i':
			cnt_inodes = TRUE;
			break;
#ifdef LSTAT
		case 'l':
			sw_follow_links = 0;
			break;
#endif
#ifdef MEMORY_BASED
		case 'o':
			sort = TRUE;
			break;
#endif
		case 's':
			sum = TRUE;
			break;
		case 't':
			sw_summary = TRUE;
			break;
		case 'q':
			quick = TRUE;
			duplicate = FALSE;
			sum = FALSE;
			cnt_inodes = FALSE;
			break;
		case 'v':
			visual = TRUE;
			break;
		case 'V':
			version++;
			break;
		default:
			err = TRUE;
		}
		if (err)
		{
			char errorMessage[75] = "[ -d ] [ -h # ] [ -i ] ";
#ifdef LSTAT
			strcat(errorMessage, "[ -l ] ");
#endif
#ifdef MEMORY_BASED
			strcat(errorMessage, "[ -o ] ");
#endif
			strcat(errorMessage, "[ -s ] [ -q ] [ -v ] [ -V ]\n");

			fprintf(stderr, "%s: %s", Program, errorMessage);
			fprintf(stderr, "	-d	count duplicate inodes\n");
			fprintf(stderr, "	-f	floating column widths\n");
			fprintf(stderr, "	-h #	height of tree to look at\n");
			fprintf(stderr, "	-i	count inodes\n");
#ifdef LSTAT
			fprintf(stderr, "	-l	Not following symbolic links\n");
#endif
#ifdef MEMORY_BASED
			fprintf(stderr, "	-o	sort directories before processing\n");
#endif
			fprintf(stderr, "	-s	include subdirectories not shown due to -h option\n");
			fprintf(stderr, "	-t	totals at the end\n");
			fprintf(stderr, "	-q	quick display, no counts\n");
			fprintf(stderr, "	-v	visual display\n");
			fprintf(stderr, "	-V	show current version\n");
			fprintf(stderr, "		(2 Vs shows specified options)\n");
			exit(-1);
		}
	}

	if (version > 0)
	{

#ifdef MEMORY_BASED
		printf("%s memory based\n", VERSION);
#else
				printf("%s disk based\n", VERSION);
#endif

		if (version > 1)
		{
			printf("Tree height:	%d\n", depth);
			if (duplicate)
				printf("Include duplicate inodes\n");
			if (cnt_inodes)
				printf("Count inodes\n");
			if (sum)
				printf("Include unseen subdirectories in totals\n");
			if (sw_summary)
				printf("Print totals at end\n");
			if (quick)
				printf("Quick display only\n");
			if (visual)
				printf("Visual tree\n");
			if (sort)
				printf("Sort directories before processing\n");
		}
	}

	/* If user didn't specify targets, inspect current directory. */

	if (optind >= argc)
	{
		user_file_list_supplied = 0;
	}
	else
	{
		user_file_list_supplied = 1;
	}

#ifdef BSD
	getcwd(topdir, sizeof(topdir)); /* find out where we are */
#else
			getcwd(topdir, sizeof(topdir)); /* find out where we are */
#endif

	/* Zero out grand totals */
	total_inodes = total_sizes = 0;
	/* Zero out sub_dirs */
	for (i = 0; i <= MAX_V_DEPTH; i++)
	{
		sub_dirs[i] = 0;
		sub_dirs_indents[i] = 0;
	}

	/* Inspect each argument */
	for (i = optind; i < argc || (!user_file_list_supplied && i == argc); i++)
	{
		cur_depth = inodes = sizes = 0;

		chdir(topdir);												/* be sure to start from the same place */
		get_data(user_file_list_supplied ? argv[i] : topdir, TRUE); /* this may change our cwd */

		total_inodes += inodes;
		total_sizes += sizes;
	}

	if (sw_summary)
	{
		printf("\n\nTotal space used: %ld\n", total_sizes);
		if (cnt_inodes)
			printf("Total inodes: %d\n", total_inodes);
	}

#ifdef HSTATS
	fflush(stdout);
	h_stats();
#endif

	// To clear the hash table
	for (int i = 0; i < position; i++)
		free(memory_array[i]);

	position = 0;

	exit(0);
} /* main */
